app/services/conversation_bot.rb

```diff
@@ -1,13 +1,297 @@
 class ConversationBot
+  NOW_TIME = /now|ahora/i.freeze
 
+  CALL_FOR_HELP = /\Ahelp|ayuda\z/i.freeze
+
+  NUMBER_STRINGS = {
+      0 => [/0/, /zero/, /cero/],
+      1 => [/1/, /one/, /uno/],
+      2 => [/2/, /two/, /dos/],
+      3 => [/3/, /three/, /tres/],
+      4 => [/4/, /four/, /cuatro/]
+  }.freeze
+
+  DONT_KNOW_FRAGMENTS = [/don't know/i, /no se/i, /no sé/i].freeze
+
+  # accepts a conversation at a certain state and a new message that has arrived
   def initialize convo, message
     @conversation = convo
+    @bot_counter = convo.bot_counter
+    @username = @conversation.username.split(' ').first
     @message = message
+    @body = @message.body.strip
+    @locale = convo.user.language
   end
 
+  # returns an appropriate response to the new message and updates the conversation
+  # we have one method for each conversation lifecycle and it returns the retries
+  # count for the conversation. Updating this on the conversation will update the
+  # conversation lifecyle.
   def response
-    # todo - bot based on lifecycle
-    'Hi, thanks for contacting us. Someone will text you momentarily to arrange a ride.'
+    if @body =~ CALL_FOR_HELP
+      stalled
+    else
+      current_lifecycle = @conversation.lifecycle
+      new_counter = send(current_lifecycle.to_sym)
+      # force update if counter doesn't change
+      @conversation.update_attributes(bot_counter: new_counter, updated_at: Time.now)
+      @response
```

Comment:
Setting instance variables like that has confused me before, like the "set_x" methods rails generators now put in controllers. Feels kind of like using globals. But also seems part of the Rails Way™, so all good.